### ISSUE TITLE
Issue after upgrading from 9.1.2 to 9.2.0.

### ISSUE THREAD

**Issue by mehimanshupatil** [ts-essentials/ts-essentials#323]
I am using type guard for a function which works flawlessly in v9.1.2 but when I upgraded to v9.2.0 it started giving error.
`A type predicate's type must be assignable to its parameter's type.`

```
import { MarkRequired } from "ts-essentials";

export enum AttachmentType {
  "IMAGE" = "IMAGE",
  "FILE" = "FILE",
  "CODE" = "CODE"
}

type AttachmentToBeUploaded =
  | {
      attachmentType: AttachmentType.IMAGE;
      id: string;
      localImageFileUrl?: string;
      localFile?: File;
      attachmentTypeSpecificFields: {
        name: string;
        type: string;
      };
    }
  | {
      attachmentType: AttachmentType.FILE;
      id: string;
      localImageFileUrl?: string;
      localFile?: File;
      attachmentTypeSpecificFields: {
        name: string;
      };
    }
  | {
      attachmentType: AttachmentType.CODE;
      id: string;
      attachmentTypeSpecificFields: {
        language: string;
      };
    };

export default function App() {
  const isAttachmentUploadable = (
    upload: AttachmentToBeUploaded
  ): upload is MarkRequired<
    Extract<
      AttachmentToBeUploaded,
      | { attachmentType: AttachmentType.IMAGE }
      | { attachmentType: AttachmentType.FILE }
    >,
    "localFile"
  > =>
    (upload.attachmentType === AttachmentType.FILE ||
      upload.attachmentType === AttachmentType.IMAGE) &&
    Boolean(upload.localFile);

  return (
    <></>
  );
}

```

reproducible codesandbox => [https://codesandbox.io/s/type-guard-error-plpnsu](https://codesandbox.io/s/type-guard-error-plpnsu)
change ts-essentials version in dependencies

**Comment by Beraliv** [comment]
Hey @mehimanshupatil!

Thank you for your report!

I checked codesandbox and think the type error makes sense now because of the changes within 9.2.0

The change was introduced here – https://github.com/ts-essentials/ts-essentials/pull/295:

```diff
- export type MarkRequired<T, RK extends keyof T> = Exclude<T, RK> & Required<Pick<T, RK>>;
+ export type MarkRequired<T, RK extends keyof T> = Omit<T, RK> & Required<Pick<T, RK>>;
```

`Exclude<T, RK>` didn't make sense for me, because it just returned `T`. But this is why it was working for you. Let me expand it for your example:

1. You have `upload: AttachmentToBeUploaded` at the beginning
2. Then you want to narrow it to 2 sybtypes from the union: with attachment type `AttachmentType.IMAGE` and `AttachmentType.FILE`
3. Once you got 2 union elements out of the whole type, you mark one field `localFile` required for both union elements
4. Previously it meant `T & Required<Pick<T, RK>>` and now it's replaced with `Omit<T, RK> & Required<Pick<T, RK>>`, it's not assignable anymore

I think we can mark it as a bug

**Comment by Beraliv** [comment]
If you have time to work on it, please feel free to create PR for it. Otherwise I will take it as soon as I have time

Thank you for understanding in advance!

**Comment by mehimanshupatil** [comment]
@Beraliv Thanks for detailed explanation 
the issue related to use of Omit with unions is mentioned here [TypeScript/issues/39556](https://github.com/microsoft/TypeScript/issues/39556)
and after updating Omit with code mentioned here https://github.com/microsoft/TypeScript/issues/39556#issuecomment-656925230
```diff
- export type MarkRequired<T, RK extends keyof T> = Omit<T, RK> & Required<Pick<T, RK>>;
+ export type MarkRequired<T, RK extends keyof T> = (T extends unknown ? Omit<T, RK> : never) & Required<Pick<T, RK>>; 
```
which works but i don't know why just changing `Omit<T, RK>` to `(T extends unknown ? Omit<T, RK> : never)` works and creating PR for same without knowing its implication for other usecases is not wise.

also should it be updated for other types with `Mark` prefix

**Comment by Beraliv** [comment]
@mehimanshupatil 

I fixed it in a PR – https://github.com/ts-essentials/ts-essentials/pull/324, can you please have a look at it? if this fix is enough and you don't have more problems, I will merge it!

Thank you for your time!

**Comment by mehimanshupatil** [comment]
@Beraliv 

tested `MarkRequired` fix works perfectly you can merge PR

**Comment by Beraliv** [comment]
I merged the changes. It will be released in ~1-2 weeks in https://github.com/ts-essentials/ts-essentials/pull/326

**Comment by Beraliv** [comment]
@mehimanshupatil thank you for your contribution again!

Released in 9.3.0 ✅
